src/fu-engine.c

```diff
@@ -1928,6 +1928,12 @@ fu_engine_update_cleanup (FuEngine *self,
 		return FALSE;
 	str = fu_device_to_string (device);
 	g_debug ("performing cleanup on %s", str);
+
+	if (fu_device_has_flag (device, FWUPD_DEVICE_FLAG_WILL_DISAPPEAR)) {
+		g_debug ("skipping cleanup due to will-disappear flag");
+		return TRUE;
+	}
+
```

Comment:
This way is more consistent with the rest of the vfuncs that are overridden when `FWUPD_DEVICE_FLAG_WILL_DISAPPEAR` is set, but the unfortunate side effect of doing it this way is `fu_plugin_runner_update_cleanup ` also won't be called.  So it's worth considering putting this block into `fu_engine_device_cleanup` instead.

Reply:
I still think we want to run cleanup even for the FWUPD_DEVICE_FLAG_WILL_DISAPPEAR case -- else it just gets a bit magic. I think we can detect the FWUPD_DEVICE_FLAG_WILL_DISAPPEAR case in the ebitdo plugin cleanup and just return TRUE there.

Reply:
The problem is that the action of opening/closing the device locker is what causes a problem in the first place.  You can't open or close a device that has already disappeared, so the device cleanup function fails.

Reply:
Aha, gotcha. I think just move this into fu_engine_device_cleanup then.